src/opencl/mod.rs

```diff
@@ -201,121 +173,174 @@ fn get_device_by_index(index: usize) -> Option<&'static Device> {
-            .len(length * std::mem::size_of::<T>())
-            .build()?;
-        buff.write(&vec![0u8]).enq()?;
+        let buff = opencl3::memory::Buffer::create(
+            &self.context,
+            CL_MEM_READ_WRITE,
+            length,
+            ptr::null_mut(),
+        )?;
+
         Ok(Buffer::<T> {
             buffer: buff,
-            _phantom: std::marker::PhantomData,
+            length,
         })
     }
 
-    pub fn create_buffer_flexible<T>(&self, max_length: usize) -> GPUResult<Buffer<T>> {
-        let mut curr = max_length;
-        let mut step = max_length / 2;
-        let mut n = 1;
-        while step > 0 && n < max_length {
-            if self.create_buffer::<T>(curr).is_ok() {
-                n = curr;
-                curr = std::cmp::min(curr + step, max_length);
-            } else {
-                curr -= step;
-            }
-            step /= 2;
-        }
-        self.create_buffer::<T>(n)
-    }
-    pub fn create_kernel(&self, name: &str, gws: usize, lws: Option<usize>) -> Kernel<'_> {
-        let mut builder = ocl::Kernel::builder();
-        builder.name(name);
-        builder.program(&self.program);
-        builder.queue(self.queue.clone());
-        builder.global_work_size([gws]);
-        if let Some(lws) = lws {
-            builder.local_work_size([lws]);
-        }
-        Kernel::<'_> { builder }
+    pub fn create_kernel(&self, name: &str, gws: usize, lws: usize) -> GPUResult<Kernel> {
+        let kernel = self
+            .kernels
+            .get(name)
+            .ok_or_else(|| GPUError::KernelNotFound(name.to_string()))?;
+        let mut builder = ExecuteKernel::new(&kernel);
+        builder.set_global_work_size(gws);
+        builder.set_local_work_size(lws);
+        Ok(Kernel {
+            builder,
+            queue: &self.queue,
+        })
     }
-}
 
-pub use ocl::OclPrm as Parameter;
+    pub fn write_from_buffer<T>(
+        &self,
+        buffer: &Buffer<T>,
+        offset: usize,
+        data: &[T],
+    ) -> GPUResult<()> {
+        assert!(offset + data.len() <= buffer.length);
+
+        let buff = buffer
+            .buffer
+            .create_sub_buffer(CL_MEM_READ_WRITE, offset, data.len())?;
+
+        self.queue
+            .enqueue_write_buffer(&buff, CL_BLOCKING, 0, data, &[])?;
+
+        Ok(())
+    }
+
+    pub fn read_into_buffer<T>(
+        &self,
+        buffer: &Buffer<T>,
+        offset: usize,
+        data: &mut [T],
+    ) -> GPUResult<()> {
+        assert!(offset + data.len() <= buffer.length);
+        let buff = buffer
+            .buffer
+            .create_sub_buffer(CL_MEM_READ_WRITE, offset, data.len())?;
+
+        self.queue
+            .enqueue_read_buffer(&buff, CL_BLOCKING, 0, data, &[])?;
+
+        Ok(())
+    }
+
+    /// Run some code in the context of the program
+    ///
+    /// On CUDA it sets the correct contexts and synchronizes the stream before returning.
```

Comment:
This comment is  a little confusing, given that no CUDA support exists right now

Reply:
Without the comment it's confusing what it is used for. Given that CUDA support is coming, I'd be in favour of just keeping it.

Reply:
yeah that's totally fine, I just wanted to mention it :)